samples/2_Cookbook/13_occupancy/occupancy.cpp

```diff
@@ -0,0 +1,157 @@
+     float eventMs = 1.0f;
+     const unsigned threadsperblock = 32;
+     const unsigned blocks = (NUM/threadsperblock)+1;
+
+     uint32_t gridSize = 0;
+     uint32_t blockSize = 0;
+      
+     if (manual){
+	blockSize = threadsperblock; 
+	gridSize  = blocks;
+	std::cout << std::endl << "Manual Configuration with block size " << blockSize << std::endl;
+     }
+     else{
+	HIP_CHECK(hipOccupancyMaxPotentialBlockSize(&gridSize, &blockSize, multiply, 0, 0));
+	std::cout << std::endl << "Automatic Configuation based on hipOccupancyMaxPotentialBlockSize " << std::endl;
+	std::cout << "Suggested blocksize is " << blockSize << ", Minimum gridsize is " << gridSize << std::endl; 
+     }
+
+     // Record the start event
+     HIP_CHECK(hipEventRecord(start, NULL));  
+
+     // Launching the Kernel from Host
+     hipLaunchKernelGGL(multiply, dim3(gridSize), dim3(blockSize), 0, 0, C, A, B, NUM);
+     
+     // Record the stop event
+     HIP_CHECK(hipEventRecord(stop, NULL));
+     HIP_CHECK(hipEventSynchronize(stop));
+         
+     HIP_CHECK(hipEventElapsedTime(&eventMs, start, stop));
+     printf("kernel Execution time = %6.3fms\n", eventMs);
+
+     //Calculate Occupancy
+     uint32_t numBlock = 0;
+     HIP_CHECK(hipOccupancyMaxActiveBlocksPerMultiprocessor(&numBlock, multiply, blockSize, 0));
+     
+     if(devProp.maxThreadsPerMultiProcessor){
+	std::cout << "Theoretical Occupancy is " << (double)numBlock* blockSize/devProp.maxThreadsPerMultiProcessor * 100 << "%" << std::endl;
+     }	
+}
+
+int main() {
+     float *A, *B, *C, *cpuC;
+     float *Ad, *Bd, *Cd;
+     int errors;
+     int i;
+
+     // initialize the input data
+     A = (float *)malloc(NUM * sizeof(float));
+     B = (float *)malloc(NUM * sizeof(float));
+     C = (float *)malloc(NUM * sizeof(float));
+     cpuC = (float *)malloc(NUM * sizeof(float));
+     
+     for(i=0; i< NUM; i++){
+	A[i] = i;
+	B[i] = i;
+     }
+    
+     // allocate the memory on the device side   
+     HIP_CHECK(hipMalloc((void**)&Ad, NUM * sizeof(float)));
+     HIP_CHECK(hipMalloc((void**)&Bd, NUM * sizeof(float)));
+     HIP_CHECK(hipMalloc((void**)&Cd, NUM * sizeof(float)));
+ 
+     // Memory transfer from host to device
+     HIP_CHECK(hipMemcpy(Ad,A,NUM * sizeof(float), hipMemcpyHostToDevice));
+     HIP_CHECK(hipMemcpy(Bd,B,NUM * sizeof(float), hipMemcpyHostToDevice));
+
+     //Kernel launch with manual/default block size
+     launchKernel(Cd, Ad, Bd, NUM, 1);
+     
+     //Kernel launch with the block size suggested by hipOccupancyMaxPotentialBlockSize 
+     launchKernel(Cd, Ad, Bd, NUM, 0);
+
+     // Memory transfer from device to host
+     HIP_CHECK(hipMemcpy(C,Cd, NUM * sizeof(float), hipMemcpyDeviceToHost));
+
+     // CPU computation
+     multiplyCPU(cpuC, A, B, NUM);
+
+     //verify the results
+     errors = 0;
+     double eps = 1.0E-6;
+     
+     for (i = 0; i < NUM; i++) {
+	  if (std::abs(C[i] - cpuC[i]) > eps) {
+		  errors++;
```

Comment:
We are effectively only checking if there are errors in the output of the 2nd kernel execution. Maybe we should either move the verification into the launchKernel function or get the kernel output into different buffers for both the kernel runs such as C0d and C1d.